extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java

```diff
@@ -0,0 +1,423 @@
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+import com.google.common.primitives.Longs;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.metamx.common.IAE;
+import com.metamx.common.ISE;
+import com.metamx.common.StringUtils;
+import com.metamx.common.logger.Logger;
+import io.druid.concurrent.Execs;
+import io.druid.query.extraction.MapLookupExtractor;
+import io.druid.server.namespace.cache.NamespaceExtractionCacheManager;
+import kafka.consumer.ConsumerConfig;
+import kafka.consumer.KafkaStream;
+import kafka.consumer.Whitelist;
+import kafka.javaapi.consumer.ConsumerConnector;
+import kafka.message.MessageAndMetadata;
+import kafka.serializer.Decoder;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.Min;
+import javax.ws.rs.GET;
+import javax.ws.rs.core.Response;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.UUID;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.regex.Pattern;
+
+@JsonTypeName("kafka")
+public class KafkaLookupExtractorFactory implements LookupExtractorFactory
+{
+  private static final Logger LOG = new Logger(KafkaLookupExtractorFactory.class);
+  static final Decoder<String> DEFAULT_STRING_DECODER = new Decoder<String>()
+  {
+    @Override
+    public String fromBytes(byte[] bytes)
+    {
+      return StringUtils.fromUtf8(bytes);
+    }
+  };
+
+  private final Object startStopLock = new Object();
+  private final ListeningExecutorService executorService;
+  private final AtomicLong doubleEventCount = new AtomicLong(0L);
+  private final NamespaceExtractionCacheManager cacheManager;
+  private final String factoryId = UUID.randomUUID().toString();
+  private final AtomicReference<Map<String, String>> mapRef = new AtomicReference<>(null);
+
+  private AtomicBoolean started = new AtomicBoolean(false);
+  private volatile ListenableFuture<?> future = null;
+
+  @JsonProperty
+  private final String kafkaTopic;
+
+  @JsonProperty
+  private final Map<String, String> kafkaProperties;
+
+  @JsonProperty
+  private final long connectTimeout;
```

Comment:
This is an extra safety valve above and beyond just the raw connection properties. If for some reason the setup process to connect to kafka is bugging out (in this class) this makes it so that the startup will fail rather than potentially loop forever.